mmv1/products/vertexai/Models.yaml

```diff
@@ -0,0 +1,419 @@
+# Copyright 2023 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+---
+name: Models
+base_url: projects/{{project}}/locations/{{region}}/models
+create_url: 'projects/{{project}}/locations/{{region}}/modelsPRE_CREATE_REPLACE_ME'
+delete_url: '{{name}}'
+self_link: '{{name}}'
+
+references:
+  guides:
+    'Official Documentation': 'https://cloud.google.com/vertex-ai/docs'
+  api: 'https://cloud.google.com/vertex-ai/docs/reference/rest/v1/projects.locations.models'
+description: |-
+  A trained machine learning Model.
+id_format: '{{name}}'
+update_verb: 'PATCH'
+exclude_import: true
+async:
+  actions: ['create', 'delete', 'update']
+  type: 'OpAsync'
+  operation:
+    base_url: '{{op_id}}'
+    path: 'name'
+    wait_ms: 1000
+  result:
+    path: 'response'
+    resource_inside_response: true
+  error:
+    path: 'error'
+    message: 'message'
+custom_code:
+  encoder: templates/terraform/encoders/vertex_ai_models.go.tmpl
+  decoder: templates/terraform/decoders/vertex_ai_models.go.tmpl
+  pre_create: templates/terraform/pre_create/vertex_ai_models.go.tmpl
+examples:
+  - name: 'vertex_ai_model_source_basic'
+    exclude_import_test: true
+    primary_resource_id: 'model'
+    vars:
+      model_id: 'test-model'
+    test_env_vars:
+      project_name: 'PROJECT_NAME'
+parameters:
+  - type: String
+    name: region
+    description: The region of the Model. eg us-central1
+    url_param_only: true
+    immutable: true
+    default_from_api: true
+properties:
+  - type: String
+    name: 'name'
+    description:
+      The name of the Model. This value may be up to 60 characters, and
+      valid characters are [a-z0-9_]. The first character cannot be a number.
+    output: true
+    # url_param_only: true
+    custom_flatten: templates/terraform/custom_flatten/vertex_ai_name_from_model.go.tmpl
+  - type: String
+    name: 'description'
+    description: Description of the Model.
+    immutable: true
```

Comment:
in this case for copy it would be output but since this resource is intending to support both copy and upload it's best to remove `output: true` as well as removing `immutable: true` since users can edit name and description on existing models:

![image](https://github.com/user-attachments/assets/3c1eed33-7275-4e63-a05c-581fed6b172a)

Reply:
Just tried this out and it works great! Just a heads-up that I need need to add back in the encoder since the `modelName` field was being used in the patch instead of `displayName`. The code would check whether to switch modelName for displayName when going into the encoder.

```hcl
	if modelName, ok := obj["modelName"].(string); ok && modelName != "" {
		obj["displayName"] = modelName
		delete(obj, "modelName")
	}
	return obj, nil
```

Reply:
I had to do quite a bit to resolve the labels diff and even then im not very satisfied with it.

Although the customize diff worked for descriptions and labels. We run into an issue where terraform applies it's own label despite the user never setting it. This causes us to see the following on apply. (This is after adding in the labels field once more since we are addressing the issue in this request thread)

```hcl
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_vertex_ai_model.name will be created
  + resource "google_vertex_ai_model" "name" {
      + artifact_uri             = (known after apply)
      + create_time              = (known after apply)
      + effective_labels         = {
          + "goog-terraform-provisioned" = "true"
        }
      + etag                     = (known after apply)
      + id                       = (known after apply)
      + model_id                 = (known after apply)
      + model_name               = (known after apply)
      + name                     = (known after apply)
      + project                  = "hc-terraform-testing"
      + region                   = (known after apply)
      + source_model             = "projects/871647908372/locations/us-central1/models/6033738282699849728"
      + supported_export_formats = (known after apply)
      + terraform_labels         = {
          + "goog-terraform-provisioned" = "true"
        }
      + trainer_pipeline         = (known after apply)
      + update_time              = (known after apply)
      + version_aliases          = (known after apply)
      + version_create_time      = (known after apply)
      + version_description      = (known after apply)
      + version_id               = (known after apply)
      + version_update_time      = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
2024-11-13T17:24:26.439-0800 [DEBUG] command: asking for input: "\nDo you want to perform these actions?"

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes
```

You'll notice that we have both effective_labels and terraform_labels. We don't apply any labels in config but these are applied automatically. Because of this we get the following error.

```hcl
│ Error: Error creating Model: googleapi: Error 400: Invalid JSON payload received. Unknown name "labels": Cannot find field.
│ Details:
│ [
│   {
│     "@type": "type.googleapis.com/google.rpc.BadRequest",
│     "fieldViolations": [
│       {
│         "description": "Invalid JSON payload received. Unknown name \"labels\": Cannot find field."
│       }
│     ]
│   }
│ ]
│ 
│   with google_vertex_ai_model.name,
│   on main.tf line 7, in resource "google_vertex_ai_model" "name":
│    7: resource "google_vertex_ai_model" "name" {
│ 
```

This is due to           `"goog-terraform-provisioned" = "true"` being include in `labels`.

I was able to resolve this by deleting `labels` from the `obj` request in `pre_create`. however we still get on plan both `effective_labels` and `terraform_labels`

This seems fine since we are expecting these fields to be filled out by api but it also seems incorrect the implementation of it all. Let me know your thoughts! @SarahFrench

Reply:
I think we should just omit labels from this PR for simplicity.

Reply:
sounds good, I'll remove the support for labels then.

Reply:
When it came to testing I found that when wanting to copy a model with an already set description led to diffs occurring on the second apply. Terraform would attempt to update the description to what is set in config (which would be null):

```hcl
2024/11/21 23:57:08 [DEBUG] [transport] [server-transport 0xc000173040] loopyWriter exiting with error: transport closed by client 
    resource_vertex_ai_model_test.go:59: Step 1/1 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # google_vertex_ai_model.model will be updated in-place
          ~ resource "google_vertex_ai_model" "model" {
              - description              = "Manually-uploaded model for testing" -> null
                id                       = "projects/1067888929963/locations/us-central1/models/2861444926426578944"
                name                     = "projects/1067888929963/locations/us-central1/models/2861444926426578944"
                # (17 unchanged attributes hidden)
        
                # (1 unchanged block hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
```

I went ahead and marked it `default_from_api: true` to resolve this. It led to all tests passing after providing it with a provisioned model_id used just for testing here. We'll be providing a model via upload once we start merging the two PRs together.

[Related Magician Output](https://github.com/GoogleCloudPlatform/magic-modules/pull/12074#issuecomment-2492591913)